src/test/java/io/confluent/examples/streams/utils/PairSerializer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.confluent.examples.streams.utils;
+
+import org.apache.kafka.common.serialization.Serializer;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.Map;
+
+public class PairSerializer<X, Y> implements Serializer<Pair<X, Y>> {
+
+  private final Serializer<X> serializerX;
+  private final Serializer<Y> serializerY;
+
+  public PairSerializer(final Serializer<X> serializerX, final Serializer<Y> serializerY) {
+    this.serializerX = serializerX;
+    this.serializerY = serializerY;
+  }
+
+  @Override
+  public void configure(final Map<String, ?> configs, final boolean isKey) {
+  }
+
+  @Override
+  public byte[] serialize(final String topic, final Pair<X, Y> pair) {
+    if (pair == null) {
+      return null;
+    }
+
+    try (
+      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      final DataOutputStream out = new DataOutputStream(baos)
+    ) {
+      final byte[] serializedX = serializerX.serialize(topic, pair.x);
+      final byte[] serializedY = serializerY.serialize(topic, pair.y);
+      short magicByte = 0;
+      if (serializedX != null) {
+        magicByte += 1;
+      }
+      if (serializedY != null) {
```

Comment:
Should this check be on `pair.y`?

Reply:
No, the magic byte logic is ultimately determined by the serialized data, not by the deserialized one.

But I think I get your point: Yeah, if the deserialized input aka `pair.y` is null, then the serialized version is likely also `null`, but this reasoning requires insider knowledge of the classes that actually do the serialization (here: `Serializer<Y>`).